src/provider/provider_coarse.c

```diff
@@ -477,16 +756,28 @@ static enum umf_result_t coarse_memory_provider_alloc(void *provider,
                        coarse_provider->alloc_size);
             }
 
+            rv = allocated_blocks_add(coarse_provider->allocated_blocks, curr);
+            if (rv) {
+                return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+            }
+
             // Try to merge new empty block with the next one.
-            block_merge_with_next(new_block, &coarse_provider->block_list);
+            new_block =
+                block_merge_with_next(coarse_provider->free_blocks, new_block,
```

Comment:
what if the merge didn't happen? we always add new_block to the free list. Please move "free_blocks_add" call to the "block_merge_with_next" func

Reply:
All is OK. There is always a new block even if the merge does not happen.